extensions/basic_auth/plugin.cc

```diff
@@ -0,0 +1,192 @@
+ */
+
+#include "extensions/basic_auth/plugin.h"
+
+#include <google/protobuf/io/zero_copy_stream_impl_lite.h>
+
+#include "absl/strings/str_cat.h"
+#include "absl/strings/str_split.h"
+#include "extensions/common/json_util.h"
+
+#ifndef NULL_PLUGIN
+
+#include "extensions/metadata_exchange/base64.h"
+
+#else
+
+#include "common/common/base64.h"
+
+namespace proxy_wasm {
+namespace null_plugin {
+namespace BasicAuth {
+namespace Plugin {
+
+PROXY_WASM_NULL_PLUGIN_REGISTRY;
+
+using Base64 = Envoy::Base64;
+
+#endif
+
+static RegisterContextFactory register_BasicAuth(
+    CONTEXT_FACTORY(PluginContext), ROOT_FACTORY(PluginRootContext));
+
+struct S {
+  std::unordered_set<std::string> requestMethods;
+  std::unordered_set<std::string> encodedCredentials;
+};
+
+using ::nlohmann::json;
+using ::Wasm::Common::JsonArrayIterate;
+using ::Wasm::Common::JsonGetField;
+std::unordered_map<std::string, S> configurationData;
+
+bool PluginRootContext::onConfigure(size_t size) {
+  // Parse configuration JSON string.
+  if (size > 0 && !configure(size)) {
+    LOG_WARN("configuration has errors, but initialzation can continue.");
+  }
+  return true;
+}
+
+bool PluginRootContext::configure(size_t configuration_size) {
+  auto configuration_data = getBufferBytes(WasmBufferType::PluginConfiguration,
+                                           0, configuration_size);
+  // Parse configuration JSON string.
+  auto result = ::Wasm::Common::JsonParse(configuration_data->view());
+  if (!result.has_value()) {
+    LOG_WARN(absl::StrCat("cannot parse plugin configuration JSON string: ",
+                          configuration_data->view()));
+    return false;
+  }
+  // j is a JsonObject holds configuration data
+  auto j = result.value();
+
+  if (!JsonArrayIterate(j, "pathPrefixes", [&](const json& prefixes) -> bool {
+        auto pathPrefix =
+            JsonGetField<std::string>(prefixes, "pathPrefix").value_or("");
+        if (pathPrefix == "") return false;
+        struct S s1;
+        if (!JsonArrayIterate(prefixes, "requestMethods",
+                              [&](const json& methods) -> bool {
+                                s1.requestMethods.insert(std::string(methods));
+                                return true;
+                              })) {
+          LOG_WARN("Failed to parse configuration for request methods.");
+          return false;
+        }
+
+        if (!JsonArrayIterate(prefixes, "credentials",
+                              [&](const json& credentials) -> bool {
+                                auto cred = std::string(credentials);
+                                s1.encodedCredentials.insert(
+                                    Base64::encode(cred.data(), cred.size()));
+                                return true;
+                              })) {
+          LOG_WARN("Failed to parse configuration for credentials.");
+          return false;
+        }
+        auto checkPath = configurationData.find(pathPrefix);
```

Comment:
c++ variable names should be snake case. Please update all variables in this change to follow the pattern.

Reply:
Also please rename this to something ended with `_iter`.

Reply:
okok updated all variables